[CapturePromotion] Preserve DebugInfo when creating project_box.

dcci opened the pull request:
The code was assigning an incorrect lexical scope, which was
responsible for an hole (and likely, wrong DI at `-Onone` in some
cases).

Fixes SR-6709.

Commit:
[CapturePromotion] Preserve DebugInfo when creating project_box.

The code was assigning an incorrect lexical scope, which was
responsible for an hole (and likely, wrong DI at `-Onone` in some
cases).

Fixes SR-6709.

dcci commented:
@swift-ci please test

adrian-prantl reviewed (approved):
This LGTM, thanks!

adrian-prantl commented on test/SILOptimizer/sr6709.swift:
```diff
@@ -0,0 +1,35 @@
+// CHECK: sil hidden @_T04null19captureStackPromoteSiycyF : $@convention(thin) () -> @owned @callee_guaranteed () -> Int {
+// CHECK: bb0:
+// CHECK:   %0 = alloc_box ${ var Int }, var, name "x", loc {{.*}}:32:7, scope 3 // users: %19, %7, %1
+// CHECK:   %1 = project_box %0 : ${ var Int }, 0, loc {{.*}}:32:7, scope 3 // users: %9, %6
```
Leaving the comments in the CHECK lines is probably just asking for trouble. I would remove them.

dcci replied:
I'll update this :)

adrian-prantl reviewed (commented):
Side-note: Swift frontend folks have already claimed the "DI" namespace for "definite initialization", better spell out debug info :-)

vedantk commented:
Nice. Just idly curious -- what are the general guidelines when deciding whether to pick SILBuilder or SILBuilderWithScope? Do you only want the former for SIL that isn't meant to be debugged?

adrian-prantl commented:
You want to use SILBuilder when you are creating new code from scratch.
You want to use SILBuilderWithScope when you are expanding a SIL instruction into a different sequence, or when you are moving SIL instructions around.

adrian-prantl commented:
When you are using SILBuilder you have to make sure to manually set a meaningful SILDebugScope.

Commit:
[CapturePromotion] Catch up with mangler changes.

Also simplify check lines and rename the test to be more
descriptive. Thanks to Adrian and Michael for pointing out.

dcci commented:
Addressed Adrian's comments. Thanks!

dcci commented:
@swift-ci please test

gottesmm commented:
@dcci can you give a more descriptive name to the test file or even better just add this to the capture promotion test?

dcci commented:
@gottesmm I'd prefer to keep this one separate. What about `capturepromotion-wrong-lexical-scope.swift` ? Happy to hear if you have another suggestion.

dcci commented:
@swift-ci please test

gottesmm commented:
@dcci Why do you want it to be separate? Couldn't you make the capture promotion test correct in this way. Then when people make changes to CapturePromotion, you get more verification for free.

dcci commented:
I don't get the "you can't make the test correct in this way".
Also, I don't see how moving to the same file of other tests results in verification for free.
Can you please elaborate?

dcci commented:
Another reason to keep this test separate is that it's very sensitive to line informations.
Imagine you add to a file with other N tests, and then you modify one of the K < N tests above this, you need to change all the check lines.

dcci commented:
I and Michael chatted about this offline a bit.
I'm going to merge this and then try to enable the verification selectively on the tests where this passes.
